### PR TITLE
Remove the EFI_NO_CONFIG_WORKING_COPY option - it is now the default.

### PR DESCRIPTION
**Pull request by racer-coder**
EDIT:  New goal:
    Any change entered via TunerStudio will now be applied immediately (modulo activeConfiguration
    stuff - TBD).  This makes for a more consistent user experience.  It also save 20,000 bytes of BSS.

### CODE REVIEW

**Comment by rusefillc** [comment]
amazing stuff!

File: firmware/console/binary/tunerstudio.cpp [hunk 1]
```diff
@@ -101,7 +101,7 @@ extern persistent_config_container_s persistentState;
  * this is a local copy of the configuration. Any changes to this copy
  * have no effect until this copy is explicitly propagated to the main working copy
  */
-persistent_config_s configWorkingCopy;
+engine_configuration_s engineConfigWorkingCopy;
```
Comment (rusefillc) [inline firmware/console/binary/tunerstudio.cpp:104]: Can we move PR comment into codebase to a larger extent?

overall my point is that PR comments are often misplaced code comments. taken to extreme in general PR description could be tiny while whatever we have to say is best place in the code

**Review by rusefillc — CHANGES_REQUESTED** [review]

File: firmware/console/binary/tunerstudio.cpp (hunk 1 above)
Comment (racer-coder) [reply firmware/console/binary/tunerstudio.cpp]: PR comment describes the delta, which is usually only/most interesting at commit time.

Comment (rusefillc) [reply firmware/console/binary/tunerstudio.cpp]: i believe following could be added into code comments:

"scatter/gather overhead so reads and writes can span both portions"
"this XXX holds a copy while YYY does ZZZ"

Comment (mck1117) [reply firmware/console/binary/tunerstudio.cpp]: @rusefillc PR has changed a bit after some discussion on Slack

Comment (rusefillc) [reply firmware/console/binary/tunerstudio.cpp]: @mck1117 need your help - no brain available today for this PR. Relying on your approval.

Comment (mck1117) [reply firmware/console/binary/tunerstudio.cpp]: I want to do some testing in actual hardware, will do tonight.

**Comment by racer-coder** [comment]
Updated version cleans up onlineApply... function.  I would like to leverage that to get rid of the notion of Burn.  Down the road we can repurpose "Burn" to mean actually flashing rather than "making some of the config live even though other parts of the config are already live"

File: firmware/console/binary/tunerstudio.h [hunk 2]
```diff
@@ -42,8 +42,6 @@ bool handlePlainCommand(TsChannelBase* tsChannel, uint8_t command);
  */
 void handleQueryCommand(TsChannelBase* tsChannel, ts_response_format_e mode);
 
-char *getWorkingPageAddr();
```
Comment (rusefillc) [inline firmware/console/binary/tunerstudio.h:45]: if it's not in header shall we make it static to emphasize?

**Comment by rusefillc** [comment]
Hold on, this is taking an unexpected turn, I wonder if applying settings on the fly is pretty much a no-no. We fatal on invalid settings, would not this change cause many settings to be invalid if done atomically one by one?

For instance you want to flip pins. If you apply settings right away you get a pin conflict on first step?

**Comment by mck1117** [comment]
> 
> 
> Hold on, this is taking an unexpected turn, I wonder if applying settings on the fly is pretty much a no-no. We fatal on invalid settings, would not this change cause many settings to be invalid if done atomically one by one?
> 
> For instance you want to flip pins. If you apply settings right away you get a pin conflict on first step?

this doesn't change the behavior for `activeConfiguration` vs `engineConfiguration` - just how they're written in to `engineConfiguration` in the first place.  We still wait for the burn button to perform pin change callbacks and things.

**Comment by rusefillc** [comment]
PS: so the proper way forward would be to add integration/unit test of the scenario I am trying to articulate as a prerequisite PR and then rebase this PR on master once it has additional coverage.

**Comment by rusefillc** [comment]
So how do we handle "oh let's me disconnect TS without burning"? And TS re-connect after the fact?

Would TS read uncommited state as truth and save it into PC config?

**Comment by racer-coder** [comment]
> Hold on, this is taking an unexpected turn, I wonder if applying settings on the fly is pretty much a no-no. We fatal on invalid settings, would not this change cause many settings to be invalid if done atomically one by one?
> 
> For instance you want to flip pins. If you apply settings right away you get a pin conflict on first step?

Maybe the code shouldn't fatal then, just put the engine in a "no run" state and wait for a valid config.

Normal engine management systems apply changes right away, unless it's a reboot-only change.  Burn shouldn't mean "use this," it should mean "save to flash."

**Comment by racer-coder** [comment]
> > Hold on, this is taking an unexpected turn, I wonder if applying settings on the fly is pretty much a no-no. We fatal on invalid settings, would not this change cause many settings to be invalid if done atomically one by one?
> > For instance you want to flip pins. If you apply settings right away you get a pin conflict on first step?
> 
> this doesn't change the behavior for `activeConfiguration` vs `engineConfiguration` - just how they're written in to `engineConfiguration` in the first place. We still wait for the burn button to perform pin change callbacks and things.

I'm changing that next!  Another 6000 bytes.  I'm making a delta structure that listeners can use to read the old values (using template and #define magic).

**Comment by rusefillc** [comment]
This is stressing me out. This PR is too big unexpectedly.

**Comment by rusefillc** [comment]
We have no automated coverage for all this. I am not available to test this all manually. This PR can only go forward as a second PR after test coverage already added into master. See coverage report.

File: firmware/console/binary/tunerstudio.h (hunk 2 above)
Comment (racer-coder) [reply firmware/console/binary/tunerstudio.h:45]: fixed

**Comment by mck1117** [comment]
> This is stressing me out. This PR is too big unexpectedly.

calm down a little bit, as discussed, I'm going to try and test this exhaustively tonight.

There are three categories of config:

1. tables, etc that exist outside of engineConfiguration
2. mostly scalar parameters, a few tables that are inside engineConfiguration
3. stuff that uses activeConfiguration, pinouts, etc

For each of the above, this PR has the following effect:

1. none
2. These parameters apply when you type them in, not when you press burn (afaik, to test on real hw tonight)
3. none

**Comment by racer-coder** [comment]
> Hold on, this is taking an unexpected turn, I wonder if applying settings on the fly is pretty much a no-no. We fatal on invalid settings, would not this change cause many settings to be invalid if done atomically one by one?
> 
> For instance you want to flip pins. If you apply settings right away you get a pin conflict on first step?

I finally have real hardware to try out.  Today, if I flip pins illegally and press Burn, all I get is a firmwareError.  It doesn't crash the box at all.  You can correct the scenario by changing the pin back, but the error doesn't clear until you reset the box.  This is a separate issue IMO - you should be able to clear errors without resetting the MCU.

So I don't see what the worry is at all - either we can handle invalid settings without crashing or we can't - if we can't that needs to be fixed whether settings happen right away or not.

**Comment by rusefillc** [comment]
let's keep this open as source of inspiration

**Comment by mck1117** [comment]
#3630

**Pull request closed without merging**